src/delagent/agent/util.c

```diff
@@ -28,148 +29,324 @@ int Verbose = 0;
 
+int verbosePrintf (const char *format, ...)
+{
+  va_list arg;
+  int done = 0;
+
+  if (Verbose)
+  {
+    va_start (arg, format);
+    done = vprintf(format, arg);
+    va_end (arg);
+  }
+  return done;
+}
+
+/**
+ * \brief PQexecCheck()
+ *
+ * simple wrapper which includes PQexec and fo_checkPQcommand
+ *
+ */
+PGresult * PQexecCheck(const char *desc, char *SQL, char *file, const int line)
+{
+  PGresult *result;
+
+  if(desc == NULL)
+  {
+    verbosePrintf("# %s:%i: %s\n", file, line, SQL);
+  }
+  else
+  {
+    verbosePrintf("# %s:%i: %s (%s)\n", file, line, desc, SQL);
+  }
+
+  result = PQexec(db_conn, SQL);
+  if (fo_checkPQcommand(db_conn, result, SQL, file, line))
+  {
+    exit(-1);
+  }
+  return result;
+}
+
+void PQexecCheckClear(const char *desc, char *SQL, char *file, const int line)
+{
+  PGresult *result;
+  result = PQexecCheck(desc, SQL, file, line);
+  PQclear(result);
+}
+
+/**
+ * \brief if this account is valid
+ *
+ * \param char *user - user name
+ * \param char *password - password
+ * \param int *user_id - will be set to the id of the user
+ * \param int *user_perm - will be set to the permission level of the user
+ *
+ * \return 1: yes, valid;
+ *         0: invalid;
+ *        -1: failure
+ */
+int authentication(char *user, char *password, int *user_id, int *user_perm)
+{
+  if (NULL == user || NULL == password)
+  {
+    return 0;
+  }
+  char SQL[MAXSQL] = {0};
+  PGresult *result;
+  char user_seed[myBUFSIZ] = {0};
+  char pass_hash_valid[41] = {0};
+  unsigned char pass_hash_actual_raw[21] = {0};
+  char pass_hash_actual[41] = {0};
+
+  /** get user_seed, user_pass on one specified user */
+  snprintf(SQL,MAXSQL,"SELECT user_seed, user_pass, user_perm, user_pk from users where user_name='%s';", user);
+  result = PQexec(db_conn, SQL);
+  if (fo_checkPQresult(db_conn, result, SQL, __FILE__, __LINE__))
+  {
+    return -1;
+  }
+  if (!PQntuples(result))
+  {
+    return 0;
+  }
+  strcpy(user_seed, PQgetvalue(result, 0, 0));
+  strcpy(pass_hash_valid, PQgetvalue(result, 0, 1));
+  *user_perm = atoi(PQgetvalue(result, 0, 2));
+  *user_id = atoi(PQgetvalue(result, 0, 3));
+  PQclear(result);
+  if (user_seed[0] && pass_hash_valid[0])
+  {
+    strcat(user_seed, password);  // get the hash code on seed+pass
+    SHA1((unsigned char *)user_seed, strlen(user_seed), pass_hash_actual_raw);
+  }
+  else
+  {
+    return -1;
+  }
+  int i = 0;
+  char temp[256] = {0};
+  for (i = 0; i < 20; i++)
+  {
+    snprintf(temp, 256, "%02x", pass_hash_actual_raw[i]);
+    strcat(pass_hash_actual, temp);
+  }
+  return strcmp(pass_hash_valid, pass_hash_actual) == 0;
+}
+
+/**
+ * \brief check if the upload can be deleted, that is the user have
+ * the permission to delete this upload
+ *
+ * \param long upload_id - upload id
+ * \param char *user_name - user name
+ *
+ * \return 1: yes, you have the needed permissions;
+ *         0: no;
+ *        -1: failure;
+ *        -2: does not exist
```

Comment:
Do you have a better Idea?

An alternative would be to use:
* `0` for *yes*
* `1` for *no* and *failure*

Then there is the problem that *does not exist* is not always a failure, but I'm currently not sure whether this differentiation is still needed.